parsl/serialize/facade.py

```diff
@@ -0,0 +1,183 @@
+            self.methods_for_code[key].enable_caching(maxsize=128)
+
+        for key in METHODS_MAP_DATA:
+            self.methods_for_data[key] = METHODS_MAP_DATA[key]()
+
+    def _list_methods(self):
+        return self.methods_for_code, self.methods_for_data
+
+    def pack_apply_message(self, func, args, kwargs,
+                           buffer_threshold=None,
+                           item_threshold=None):
+        """Serialize and pack function and parameters
+
+        Parameters
+        ----------
+
+        func: Function
+            A function to ship
+
+        args: Tuple/list of objects
+            positional parameters as a list
+
+        kwargs: Dict
+            Dict containing named parameters
+
+        buffer_threshold: Ignored
+            This kwarg is provided only to match the interface to ipyparallel.serialize
+
+        item_threshold: Ignored
+            This kwarg is provided only to match the interface to ipyparallel.serialize
+        """
+        b_func = self.serialize(func)
+        b_args = self.serialize(args)
+        b_kwargs = self.serialize(kwargs)
+        packed_buffer = self.pack_buffers([b_func, b_args, b_kwargs])
+        return packed_buffer
+
+    def unpack_apply_message(self, packed_buffer, user_ns=None, copy=False):
+        """ Unpack and deserialize function and parameters
+
+        """
+        return [self.deserialize(buf) for buf in self.unpack_buffers(packed_buffer)]
+
+    def serialize(self, obj, buffer_threshold=1e6):
+        """ Try available serialization methods one at a time
+
+        If all serialization methods fail we raise a TypeError. Ideally we should
+        reraise the exception from the methods we tried which might have more
+        useful info into why the object cannot be serialized
+        TODO ^
+        """
+        serialized = None
+        serialized_flag = False
+        kind = None
+        if callable(obj):
+            kind = 'callable'
+            for method in self.methods_for_code.values():
+                try:
+                    serialized = method.serialize(obj)
+                    # We attempt a deserialization to make sure both work.
+                    method.deserialize(serialized)
+                except Exception:
+                    logger.exception(f"Serialization method: {method} did not work")
+                    continue
+                else:
+                    serialized_flag = True
+                    break
+        else:
+            kind = 'data'
+            for method in self.methods_for_data.values():
+                try:
+                    serialized = method.serialize(obj)
+                except Exception:
+                    logger.exception(f"Serialization method {method} did not work")
+                    continue
+                else:
+                    serialized_flag = True
+                    break
+
+        if serialized_flag is False:
+            # TODO : Replace with a SerializationError
```

Comment:
TODO!

Reply:
Replaced with raising the last exception raised by the serializer methods.